### PR TITLE
React: Upgrade to version 18

### PR DESCRIPTION
**Pull request by kaydelaney**
**What this PR does / why we need it**:
Upgrades react and related packages.

**Special notes for your reviewer**:
I've tried to structure this PR in a way that's a little easier to consume than if it were just a single commit, so for anyone reviewing, I'd suggest going one commit at a time.

***Changes***:
Most of the changes are type-related, owing to `React.FC`/`React.FunctionComponent` no longer implicitly adding a `children` prop to components. Several usages of `useCallback` were updated to specify the callback param args.
 
Since enzyme doesn't support react 18, I've also removed the old enzyme slate tests and re-wrote them as e2e tests. Some other enzyme tests were converted to rtl, but I also just commented some out as a to-do.

The `Strict` wrapper around the app was removed as it breaks `react-aria` unfortunately. Hopefully it can be added back in the near future.

There are still a lot of failing tests, mostly related to React complaining about things not being wrapped in `act`. It seems like InlineSVG might be the culprit, but I need to investigate more.

### CODE REVIEW

**Comment by Clarity-89** [comment]
> Most of the changes are type-related, owing to React.FC/React.FunctionComponent no longer implicitly adding a children prop to components. Several usages of useCallback were updated to specify the callback param args.

Wonder if these changes could be done before updating to React 18, in a separate PR, so this one has less changes, or do they depend on React 18?

**Comment by kaydelaney** [comment]
@Clarity-89 I could do that, but I've contained the type changes mentioned more or less to the first commit, so you can just ignore that one if it makes things simpler

File: packages/grafana-runtime/package.json [hunk 1]
```diff
@@ -49,31 +49,31 @@
...
     "rollup-plugin-sourcemaps": "0.6.3",
     "rollup-plugin-terser": "7.0.2",
-    "typescript": "4.8.2"
+    "typescript": "4.8.3"
   },
   "peerDependencies": {
-    "react": "17.0.2",
-    "react-dom": "17.0.2"
+    "react": "18.2.0",
+    "react-dom": "18.2.0"
```
Comment (jackw) [inline packages/grafana-runtime/package.json:77]: I have a feeling this is a mistake on my part. It should likely follow the same semver ranges as the other packages:

```suggestion
    "react": "^16.8.0 || ^17.0.0 || ^18.0.0",
    "react-dom": "^16.8.0 || ^17.0.0 || ^18.0.0"
```

**Review by jackw — COMMENTED** [review]
Amazing progress on this @kaydelaney . I can see there are quite a few bumps to dependencies in toolkit and there is also changes to the `jest.config.js`. Toolkit also has a jest config [here](https://github.com/grafana/grafana/blob/main/packages/grafana-toolkit/src/config/jest.plugin.config.ts) which might require some updates. Maybe once this PR is out of draft we could try testing these changes against some community plugins with tests? WDYT?

**Comment by kaydelaney** [comment]
> Maybe once this PR is out of draft we could try testing these changes against some community plugins with tests? WDYT?

I think that's a very good idea

**Comment by kaydelaney** [comment]
This PR solves some e2e flakiness that the upgrade introduced for whatever reason: https://github.com/grafana/grafana/pull/55490

**Pull request closed without merging**